import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.Set;
import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.TypeElement;
import toolbox.JavacTask;
import toolbox.Task;
import toolbox.Task.Mode;
import toolbox.Task.OutputKind;
import toolbox.TestRunner;
import toolbox.ToolBox;

public class CrashEmptyEnumConstructorTest extends TestRunner {
    protected ToolBox tb;
    CrashEmptyEnumConstructorTest() {
        super(System.err);
        tb = new ToolBox();
    }
    public static void main(String... args) throws Exception {
        new CrashEmptyEnumConstructorTest().runTests();
    }
    protected void runTests() throws Exception {
        runTests((m) -> new Object[] {Paths.get(m.getName())});
    }
    Path[] findJavaFiles(Path... paths) throws IOException {
        return tb.findJavaFiles(paths);
    }
    @Test
    public void testEmptyEnumConstructor(Path base) throws Exception {
        Path src = base.resolve("src");
        Path r = src.resolve("E");
        Path classes = base.resolve("classes");
        Files.createDirectories(classes);
        tb.writeJavaFiles(r, """
                enum E {
                    ONE("");
                    E(String one);
                }
                """);
        List<String> expected = List.of("E.java:3: error: method E(String) in E is missing a method body, or should be declared abstract", "    E(String one);", "    ^", "1 error");
        List<String> log = new JavacTask(tb).options("-processor", SimpleProcessor.class.getName()).files(findJavaFiles(src)).outdir(classes).run(Task.Expect.FAIL).writeAll().getOutputLines(Task.OutputKind.DIRECT);
        if (log.size() != expected.size()) {
            throw new AssertionError("Unexpected output: " + log);
        } else {
            for (int i = 0; i < expected.size(); i++) {
                if (!log.get(i).contains(expected.get(i))) {
                    throw new AssertionError("Unexpected output: " + log);
                }
            }
        }
    }
    @SupportedAnnotationTypes("*")
    public static final class SimpleProcessor extends AbstractProcessor {
        @Override
        public SourceVersion getSupportedSourceVersion() {
            return SourceVersion.latestSupported();
        }
        @Override
        public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
            return false;
        }
    }
}
